cudaextender/src/ungapped_xdrop.cu

```diff
@@ -0,0 +1,227 @@
+        // TODO- Make async
+        const int32_t num_scored_segment_pairs = get_value_from_device(d_done_.data() + curr_num_pairs - 1, stream_);
+        if (num_scored_segment_pairs > 0)
+        {
+            compress_output<<<1024, 1024, 0, stream_>>>(d_done_.data(),
+                                                        seed_pair_start,
+                                                        d_scored_segment_pairs,
+                                                        d_tmp_ssp_.data(),
+                                                        curr_num_pairs); // TODO- Need configurability for kernel?
+            thrust::stable_sort(thrust::cuda::par(allocator_).on(stream_),
+                                d_tmp_ssp_.begin(),
+                                d_tmp_ssp_.begin() + num_scored_segment_pairs,
+                                scored_segment_pair_comp());
+            GW_CU_CHECK_ERR(cub::DeviceSelect::Unique(d_temp_storage_cub_.data(),
+                                                      cub_storage_bytes,
+                                                      d_tmp_ssp_.data(),
+                                                      d_scored_segment_pairs + total_scored_segment_pairs_,
+                                                      d_num_scored_segment_pairs,
+                                                      num_scored_segment_pairs,
+                                                      stream_))
+            total_scored_segment_pairs_ += get_value_from_device(d_num_scored_segment_pairs, stream_);
+        }
+    }
+    auto t2 = std::chrono::high_resolution_clock::now();
+
+    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
+
+    std::cout << "Time: " << duration << std::endl;
+
+    set_device_value_async(d_num_scored_segment_pairs, &total_scored_segment_pairs_, stream_);
+    return success;
+}
+
+StatusType UngappedXDrop::extend_async(const char* h_query, const int32_t& query_length,
+                                       const char* h_target, const int32_t& target_length,
+                                       const int32_t& score_threshold,
+                                       const std::vector<SeedPair>& h_seed_pairs)
+{
+    // Reset the extender if it was used before in this mode
+    reset();
+    // Set host pointer mode on
+    host_ptr_api_mode_ = true;
+    // Allocate space for query and target sequences
+    d_query_  = device_buffer<char>(query_length, allocator_, stream_);
+    d_target_ = device_buffer<char>(target_length, allocator_, stream_);
+    // Allocate space for SeedPair input
+    d_seed_pairs_ = device_buffer<SeedPair>(h_seed_pairs.size(), allocator_, stream_);
+    // Allocate space for ScoredSegmentPair output
+    d_ssp_ = device_buffer<ScoredSegmentPair>(h_seed_pairs.size(), allocator_, stream_);
+    GW_CU_CHECK_ERR(cudaMalloc((void**)&d_num_ssp_, sizeof(int32_t)))
```

Comment:
why can't this be a `device_buffer`?

Reply:
I was wondering whether the overhead of a device_buffer would make it worth it for a single int.

Reply:
good point, but it would. The overhead of `device_buffer` is lower than a `cudaMalloc` call.

Reply:
Done.